Additional claims enhancements (#15491)
aspnet/AspNetCore.Docs: aspnetcore/security/authentication/social/additional-claims.md

* Limit the number and size of user claims for request processing to only what the app requires.
* Use a custom <xref:Microsoft.AspNetCore.Authentication.Cookies.ITicketStore> for the Cookie Authentication Middleware's <xref:Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationOptions.SessionStore> to store identity across requests. Preserve large quantities of identity information on the server while only sending a small session identifier key to the client.

> Claims are added to the user on first registration, not on sign in. If additional claims are enabled in an app after a user registers to use the app, delete the user and have them register again.

Reviewer: ```suggestion
Claims are added to the user on first registration, not on sign in. If additional claims are enabled in an app after a user registers to use the app they won't be available.
```
Deleting and recreating a user is only viable in a dev environment. @HaoK how do you update external user claims at signin?

Reviewer 2: You can probably just call ReplaceClaim on the user for claims stored in the identity database https://github.com/aspnet/AspNetCore/blob/master/src/Identity/Extensions.Core/src/UserManager.cs#L1107

Reviewer 2: you would have to also call RefreshSign in to force a new cookie

Author: We can work on the text directly in the suggestion :point_up: until it's correct and then add it.

Reviewer: It needs to say where to call these.

Author: @HaoK I checked our docs hoping that I could cross-link to content explaining the pattern. AFAICT, we don't have general guidance for this scenario.

The silent re-sign in with `RefreshSignInAsync` isn't explicitly covered, but that's easy to grok from samples where a user updates their profile, changes their password, etc.

Replacing the claim:

1. Isn't "adding" performed with `AddClaimAsync`? The scenario that I envision thus far is that there is a new claim that the dev wants to take from their social auth provider.
1. What would you like to show for adding/changing claims? ... a `IUserClaimsPrincipalFactory` that runs on users?
1. What if the dev prefers instead a utility approach that they can run across all of the users at once? ... also something that I don't think docs show (but I admit that I'm using search to look for things and not reading :eyes: every doc we have).

Author: @HaoK ... I need additional guidance ...

Paraphrased ...

> You can call `ReplaceClaimAsync` on the user for claims stored in the identity database followed by `RefreshSignInAsync`.

I'm not sure who "you" is in this context or why it's a "replacement" when the claim is new for the app.

I envision a scenario where users are already registered. The dev wants to add a new claim provided by the social auth provider to their app. I wondered if the dev should create an `IUserClaimsPrincipalFactory` that calls `AddClaimAsync` if the claim isn't already present.

It's best to ask before going down that road with a test app because if it's the wrong approach, it will save some 🕐 ... thus 💰.

Reviewer 2: So the intent of refresh sign in, is that you specify what user you wish the regenerate the claims for, it doesn't have any safeguard to enforce that you are refreshing the user you originally generated the cookie from.  Basically it will recreate the claims for that user and call sign in with them but also will attempt to preserve the original metadata stored in the cookie (which preserves things like remember me)

See https://github.com/aspnet/AspNetCore/blob/master/src/Identity/Core/src/SignInManager.cs#L169

Reviewer 2: You refers to the developer who wants to update a claim for the user, the app could on subsequent external logins call replace claim with the updated value of a claim.  Does that make sense?  It shouldn't be necessary for them to have a custom principal factory for this.

* Limit the number and size of user claims for request processing to only what the app requires.
* Use a custom <xref:Microsoft.AspNetCore.Authentication.Cookies.ITicketStore> for the Cookie Authentication Middleware's <xref:Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationOptions.SessionStore> to store identity across requests. Preserve large quantities of identity information on the server while only sending a small session identifier key to the client.

> Claims are added to the user on first registration, not on sign in. If additional claims are enabled in an app after a user registers to use the app, delete the user and have them register again.

Author: ```suggestion
Claims are added to the user on first registration, not on sign in. If additional claims are enabled in an app after a user registers to use the app:

* Call [UserManager.ReplaceClaimAsync](xref:Microsoft.AspNetCore.Identity.UserManager*) on the user for claims stored in the identity database.
* Call [SignInManager.RefreshSignInAsync](xref:Microsoft.AspNetCore.Identity.SignInManager*) to force the generation of a new authentication cookie.
```
